flare/dft_interface/vasp_util.py

```diff
@@ -0,0 +1,128 @@
+from pymatgen.io.vasp.inputs import Poscar
+from pymatgen.io.vasp.outputs import Vasprun
+from subprocess import call
+import numpy as np
+from flare.struc import Structure
+from typing import List, Union
+from json import dump, load
+from flare.util import NumpyEncoder
+import os, shutil
+
+def check_vasprun(vasprun: Union[str, Vasprun]) -> Vasprun:
+    """
+    Helper utility to take a vasprun file name or Vasprun object
+    and return a vasprun object.
+    :param vasprun: vasprun filename or object
+    """
+    if type(vasprun) == str:
+        return Vasprun(vasprun)
+    elif type(vasprun) == Vasprun:
+        return vasprun
+    else:
+        raise ValueError('Vasprun argument is not a string or Vasprun instance!')
+
+def run_dft(calc_dir: str, dft_loc: str,
+    structure: Structure=None, en: bool=False, vasp_cmd="{}"):
+    """
+    Run a VASP calculation.
+    :param calc_dir: Name of directory to perform calculation in
+    :param dft_loc: Name of VASP command (i.e. executable location)
+    :param structure: flare structure object
+    :param en: whether to return the final energy along with the forces
+    :param vasp_cmd: Command to run VASP (leave a formatter "{}" to insert dft_loc);
+        this can be used to specify mpirun, srun, etc.
+
+    Returns:
+        forces on each atom (and energy if en=True)
+    """
+    
+    currdir = os.getcwd()
+    if structure:
+        edit_dft_input_positions("POSCAR", structure)
+    try:
+        if currdir != calc_dir:
+            os.chdir(calc_dir)
+        call(vasp_cmd.format(dft_loc).split())
+
+        if en:
+            parse_func = parse_dft_forces_and_energy
+        else:
+            parse_func = parse_dft_forces
+
+        try:
+            forces = parse_func("vasprun.xml")
+        except FileNotFoundError:
+            raise FileNotFoundError("""Could not load vasprun.xml. The calculation may not have finished.
+                                        Current directory is %s""" % os.getcwd())
+
+        os.chdir(currdir)
+        return forces
+
+    except Exception as e:
+        os.chdir(currdir)
+        raise e
+    
+
+def dft_input_to_structure(poscar: str):
+    """
+    Parse the DFT input in a directory.
+    :param vasp_input: directory of vasp input
+    """
+    return Structure.from_pmg_structure(Poscar.from_file(poscar).structure)
```

Comment:
Can you ensure that this returns the cartesian coordinates every time since that's the FLARE convention?

Reply:
This is addressed by the unit test for the from_pmg_structure function, but I also will address it in the `test_structure_parsing` function when I assert the correct coordinate parsing

Reply:
Actually, `test_struc.test_from_pmg_structure` doesn't guarantee the cartesian coordinates are read because it sets the unit cell to np.eye(3). So I copy-pasted the test and changed the lattice parameters so that it will fail if `from_pmg_structure` returns fractional coordinates.